networking_cisco/plugins/cisco/device_manager/plugging_drivers/aci_vlan_trunking_driver.py

```diff
@@ -0,0 +1,255 @@
+                port_db.hosting_info.segmentation_id)
+        is_external = (port_db.device_owner == DEVICE_OWNER_ROUTER_GW)
+        hosting_info['physical_interface'] = self._get_interface_info(
+            hosting_device['id'], port_db.network_id, is_external)
+        ext_dict = self._get_external_network_dict(context, port_db)
+        if not is_external:
+            hosting_info['host_pool_cidr'] = ext_dict['host_pool_cidr']
+            hosting_info['next_hop'] = ext_dict['next_hop']
+            hosting_info['gateway_ip'] = ext_dict['gateway']
+
+    def allocate_hosting_port(self, context, router_id, port_db, network_type,
+                              hosting_device_id):
+        """Get the VLAN and port for this hosting device
+
+        The VLAN used between the APIC and the external router is stored
+        by the APIC driver.  This calls into the APIC driver to first get
+        the ACI VRF information associated with this port, then uses that
+        to look up the VLAN to use for this port to the external router
+        (kept as part of the L3 Out policy in ACI).
+        """
+        # If this is a router interface, the VLAN comes from APIC.
+        # If it's the gateway, the VLAN comes from the segment ID
+        if port_db.get('device_owner') == DEVICE_OWNER_ROUTER_GW:
+            return super(AciVLANTrunkingPlugDriver,
+                         self).allocate_hosting_port(
+                             context, router_id,
+                             port_db, network_type, hosting_device_id
+                         )
+
+        # shouldn't happen, but just in case
+        if port_db.get('device_owner') != DEVICE_OWNER_ROUTER_INTF:
+            return None
+
+        # get the external network that this port connects to.
+        # if there isn't an external gateway yet on the router,
+        # then don't allocate a port
+
+        router = self.l3_plugin.get_router(context, router_id)
+        network_id = router.get('external_gateway_info', {}).get('network_id')
+        if network_id is None:
+            return None
+
+        networks = self._core_plugin.get_networks(
+            context.elevated(), {'id': [network_id]})
+        l3out_network = networks[0]
+        # For VLAN apic driver provides VLAN tag
+        kwargs = self.get_vrf_context(context, router_id, port_db)
+        details = self.apic_driver.get_vrf_details(context, **kwargs)
+        if details is None:
+            LOG.debug('aci_vlan_trunking_driver: No vrf_details')
+            return
+        vrf_name = details.get('vrf_name')
+        vrf_tenant = details.get('vrf_tenant')
+        allocated_vlan = (None if self.apic_driver is None
+            else self.apic_driver.l3out_vlan_alloc.get_vlan_allocated(
+                l3out_network['name'], vrf_name, vrf_tenant=vrf_tenant))
+        if allocated_vlan is None:
+            if vrf_tenant is None or vrf_tenant == '':
+                # TODO(tbachman): I can't remember why this is here
```

Comment:
I'll re-investigate this. I seem to remember this was here for the global routers, but I should verify that this is still needed.